### ISSUE TITLE
Subclassing InfinityModel provides no pattern to replace route._afterInfinityModel

### ISSUE THREAD

**Issue by johncalvinyoung** [ember-infinity/ember-infinity#292]
So in the new best practices pattern for bound parameters (and therefore keyset/cursor pagination), we're supposed to subclass `InfinityModel` and set up the collection with our subclass that will manage injecting the latest state for each request. I've figured out how to handle the actual after key, but correctly setting _canLoadMore as in the example requires action in the `_afterInfinityModel` hook, and there's no good way to package up that logic within the new subclass.

Example implementation:

`ListItemRoute`:
```
export default ListRoute.extend(InfinityRoute, {
  model(params) {
    return this.infinityModel('list-item', {
      sort: get(this, 'sort'),
      filter: get(this, 'filter'),
      current: get(this, 'currentDate')
    }, ListFetcher);
  },

  afterInfinityModel(listItems, infinityModel) {
    infinityModel.set('canLoadMore', listItems.get('length') > 0);
  }
}
```

`ListFetcher` (subclass of `InfinityModel`)
```
import Ember from 'ember';
import InfinityModel from 'ember-infinity/lib/infinity-model';

const { computed, get, isPresent } = Ember;

export default InfinityModel.extend({
  after: computed('content.[]', function() {
    return get(this, 'content.lastObject.after');
  }),
  pageParam: null,
  perPageParams: null,

  // need to set an after hook here, to update canLoadMore/_canLoadMore on the latest batch request.

  buildParams(increment) {
    const params = this._super(increment),
      after = get(this, 'after');

    if(isPresent(after)) {
      params['after'] = after;
    }
    return params;
  }
});

```

As you can see, there's a gap here, and it's one that I think makes sense to fix: the hook may not be/probably shouldn't be generic for all infinityModels in the route, and it'd just be cleaner if we can package that sort of logic within the InfinityModel itself. Why not make the canonical 'home' of the hook within the model, and only make it fall back to `route._afterInfinityModel` through there. Logic also applies to `.infinityModelLoaded`, though I don't personally use it.

Alternately/additionally, the concept of 'most recent page/batch loaded' is probably a useful one to populate within the `InfinityModel`. I can think of several ways that would be very useful in the non-numeric-paged world.

I can come up with a PR if everyone agrees on an approach.

**Comment by johncalvinyoung** [comment]
I can come up with a PR if everyone can agree on an approach.

**Comment by johncalvinyoung** [comment]
I’m looking at it, but it’s definitely a bit of a bear to make backwards compatible with the existing pattern.

**Comment by johncalvinyoung** [comment]
I've got a branch in progress, but I've run into a number of other interesting things as I've implemented a complex new feature on top my keyset-paginated app using `1.0.0-beta.6`.

- it would be good to let `InfinityModel` once instantiated have a link to the `infinity` service, so it can automatically request additional pages of results in certain scenarios.
- `afterInfinityModel` and `infinityModelLoaded` always presume that loading is on the end of the set. It'd be good to pass direction (and track `canLoadMore` separately) for 'up' as well as 'down', for the bidirectlonal-loading case. Maybe there are no more results 'after' the current point, but there might be some before.

Anyone have thoughts on the above? Some of this would affect interfaces somewhat...
